scripts/check-markdown-links.ts

```diff
@@ -0,0 +1,108 @@
+import { sh } from './utils';
+import path from 'node:path';
+import fs from 'node:fs';
+import fetch from 'node-fetch';
+interface FilesAndLinks {
+  [key: string]: Array<string>;
+}
+
+const appDir = path.resolve(__dirname, '..');
+const fileExtensions = ['.md', '.mdx'];
+const outputFileName = `${appDir}/brokenLinkReport.html`;
+let foundFiles: string | undefined;
+const urlsToCheck: FilesAndLinks = {};
+
+// Construct our find command:
+const findCommand = (): string => {
+  const commandBase = `find ${appDir} `;
+  const commandArgs = fileExtensions
+    .map((ext) => {
+      return `-type f -name '*${ext}' -not -path '*/node_modules/*' -not -path '*/.github/*'`;
+    })
+    .join(` -or `);
+
+  return commandBase.concat(commandArgs);
+};
+
+// Prevents us trying to call localhost urls.
+const isValidUrl = (url: string): boolean => {
+  let urlHolder: URL;
+  try {
+    urlHolder = new URL(url);
+  } catch (error) {
+    console.log(
+      `${url} could not be turned into a proper URL. Maybe explicitly exclude this type of thing?`
+    );
+    return false;
+  }
+  if (urlHolder.hostname.includes('localhost')) {
+    return false;
+  }
+  return true;
+};
+
+// Extract links from a particular md/mdx file and place into predefined object:
+const findLinks = (filePath: string, regex: RegExp): void => {
+  try {
+    const fileContent = fs.readFileSync(filePath, 'utf8');
+    const matches: RegExpMatchArray | null = fileContent.match(regex);
+    if (matches) {
+      (urlsToCheck as FilesAndLinks)[filePath] = matches.filter((match) => {
+        if (isValidUrl(match)) {
+          return match;
+        }
+      });
+    }
+  } catch (error) {
+    console.log(
+      `An error occurred when parsing this file: ${filePath}. \nHere's the error:\n ${error}`
+    );
+  }
+};
+
+// Use node-fetch to query urls and return a status code:
+const curl = (file: string, url: string): void => {
+  const spliceFrom = file.split('/').indexOf('design-system');
+  const filePath: string = file.split('/').splice(spliceFrom).join('/');
+  fetch(url).then((res) => {
+    if (res.status !== 200) {
+      const content = `<div><h4>Filepath: <a href="${file}" target="_blank">${filePath}</a></h4><div style="margin-left:.5rem"><a href="${url}" target="_blank">${url}</a> <p>Code: ${res.status}</p></div><br/></div>\n`;
+      fs.writeFile(outputFileName, content, { flag: 'a+' }, (err) => {
```

Comment:
I noticed that `fs.writeFile` is used to append results for each detected broken link. Is there a risk that broken links might be found concurrently and lead to race conditions in the write operation?

Reply:
@tamara-corbalt I think changes to the code get rid of this concern. 
1. File writing is now synchronous
2. We only write to the file in two places now: prior to the url requests to initialize the output file and once when we are done looping through the urls. That's the purpose of that 'lastItem' boolean.